libptable/ptable.c

```diff
@@ -0,0 +1,211 @@
+
+
+#if PTABLE_CHECKSUM
+static uint32_t ptable_crc32(uint32_t crc, const void *data, size_t len)
+{
+	int i;
+
+	while (len--) {
+		crc ^= *(char *)data++;
+		for (i = 0; i < 8; i++) {
+			crc = (crc >> 1) ^ ((crc & 1) ? 0xedb88320 : 0);
+		}
+	}
+
+	return crc;
+}
+#endif
+
+
+static int ptable_verifyPartition(const ptable_t *ptable, const ptable_part_t *part, const ptable_memctx_t *memctx)
+{
+	const ptable_part_t *p;
+	int i;
+
+#if PTABLE_CHECKSUM
+	/* Verify partition checksum */
+	if (part->crc != ptable_crc32(0, part, sizeof(ptable_part_t) - sizeof(uint32_t))) {
+		return -1;
+	}
+#endif
+
+	/* Verify offset and size */
+	if (part->size == 0) {
+		return -1;
+	}
+
+	if (part->size % memctx->blksz) {
+		return -1;
+	}
+
+	if (part->offset % memctx->blksz) {
+		return -1;
+	}
+
+	if (part->offset + part->size > memctx->memsz) {
+		return -1;
+	}
+
+	/* Verify partition type */
+	switch (part->type) {
+		case ptable_raw:
+		case ptable_jffs2:
+		case ptable_meterfs:
+			break;
+
+		default:
+			return -1;
+	}
+
+	/* Verify partition name */
+	for (i = 0; i < sizeof(part->name); i++) {
+		if (!isalnum(part->name[i])) {
+			break;
+		}
+	}
+
+	if ((i == 0) || (i >= sizeof(part->name)) || (part->name[i] != '\0')) {
+		return -1;
+	}
+
+	/* Compare against previous partitions */
+	for (p = ptable->parts; p != part; p++) {
+		/* Check for range overlap */
+		if ((part->offset <= p->offset + p->size - 1) && (part->offset + part->size - 1 >= p->offset)) {
+			return -1;
+		}
+
+		/* Check for name duplicate */
+		if (strcmp((const char *)part->name, (const char *)p->name) == 0) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+static int ptable_verify(const ptable_t *ptable, const ptable_memctx_t *memctx)
+{
+	int i;
+
+#if PTABLE_CHECKSUM
+	/* Verify header checksum */
+	if (ptable->crc != ptable_crc32(0, ptable, sizeof(ptable_t) - sizeof(uint32_t))) {
+		return -1;
+	}
+#endif
+
+	/* Verify number of partitions */
+	if (ptable->count * sizeof(ptable_part_t) > memctx->secsz - sizeof(ptable_t) - sizeof(ptable_magic)) {
+		return -1;
+	}
+
+	/* Verify magic bytes */
+	if (memcmp((uint8_t *)ptable + sizeof(ptable_t) + ptable->count * sizeof(ptable_part_t), ptable_magic, sizeof(ptable_magic)) != 0) {
+		return -1;
+	}
+
+	/* Verify partitions */
+	for (i = 0; i < ptable->count; i++) {
+		if (ptable_verifyPartition(ptable, ptable->parts + i, memctx) < 0) {
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+
+int ptable_read(ptable_t *ptable, const ptable_memctx_t *memctx)
+{
+	int i;
+
+	/* Find first valid partition table */
+	for (i = 0; i < PTABLE_NCOPIES; i++) {
+		/* Read sector from memory */
+		if (memctx->read(memctx->memsz - (i + 1) * memctx->blksz, ptable, memctx->secsz) != memctx->secsz) {
+			continue;
+		}
+
+		/* Verify partition table */
+		if (ptable_verify(ptable, memctx) < 0) {
+			continue;
+		}
+
+		return 0;
+	}
+
+	return -ENOENT;
+}
+
+
+int ptable_write(ptable_t *ptable, const ptable_memctx_t *memctx)
+{
+	int i;
+
+#if PTABLE_CHECKSUM
```

Comment:
I think we can cheat a little and write CRC anyway - just don't verify it

Reply:
I agree. `PTABLE_CHECKSUM` is now checked only for partition table verification.